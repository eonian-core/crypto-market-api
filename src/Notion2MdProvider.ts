import { Client } from "@notionhq/client";
import { NotionToMarkdown } from 'notion-to-md'
import type { ListBlockChildrenResponseResults, MdBlock, MdStringObject } from "notion-to-md/build/types";

export class Notion2MdProvider {
    private readonly notion: Client
    private readonly n2m: NotionToMarkdown

    constructor() {
        this.notion = new Client({})
        this.n2m = new NotionToMarkdown({
            notionClient: this.notion
        })
    }

    /**
     * Converts list of Notion Blocks to Markdown string
     * @param {ListBlockChildrenResponseResults | undefined} blocks - List of notion blocks
     * @param {number} totalPage - Retrieve block children request number, page_size Maximum = totalPage * 100
     * @param {MdBlock[]} mdBlocks - Defines max depth of nesting
     * @param {number} nestingLevel - Defines max depth of nesting
     * @returns {MdStringObject} - Returns markdown string with child pages separated
     */
    async toMd(blocks?: ListBlockChildrenResponseResults, totalPage?: number | null, mdBlocks?: MdBlock[], pageIdentifier?: string, nestingLevel?: number): Promise<MdStringObject> {
        const result = await this.toMdBlocks(blocks, totalPage, mdBlocks)
        return this.mdBlocksToString(result, pageIdentifier, nestingLevel)
    }

    /**
     * Converts list of Notion Blocks to Markdown Blocks
     * @param {ListBlockChildrenResponseResults | undefined} blocks - List of notion blocks
     * @param {number} totalPage - Retrieve block children request number, page_size Maximum = totalPage * 100
     * @param {MdBlock[]} mdBlocks - Defines max depth of nesting
     * @returns {Promise<MdBlock[]>} - Array of markdown blocks with their children
     */
    async toMdBlocks(blocks?: ListBlockChildrenResponseResults, totalPage?: number | null, mdBlocks?: MdBlock[]): Promise<MdBlock[]> {
        return await this.n2m.blocksToMarkdown(blocks, totalPage, mdBlocks)
    }

    /**
     * Converts Markdown Blocks to string
     * @param {MdBlock[]} mdBlocks - Array of markdown blocks
     * @param {number} nestingLevel - Defines max depth of nesting
     * @returns {MdStringObject} - Returns markdown string with child pages separated
     */
    mdBlocksToString(mdBlocks?: MdBlock[], pageIdentifier?: string, nestingLevel?: number): MdStringObject {
        return this.n2m.toMarkdownString(mdBlocks, pageIdentifier, nestingLevel)
    }
}