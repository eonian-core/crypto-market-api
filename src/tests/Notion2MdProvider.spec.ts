import {describe, expect, beforeEach, it} from '@jest/globals';
import { Notion2MdProvider } from '../Notion2MdProvider';

describe('Notion2MdProvider', () => {
  let provider: Notion2MdProvider;

  beforeEach(() => {
    provider = new Notion2MdProvider();
  });

  describe('toMd method', () => {
    it('should convert blocks to markdown correctly', async () => {
      const result = await provider.toMd(blocks as any);
      
      expect(result).toMatchSnapshot();
    });
  });

});

const blocks = [
  {
    "object": "block",
    "paragraph": {
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "text": {
            "content": "hello ",
            "link": undefined,
          },
          "plain_text": "hello ",
          "type": "text",
        },
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": true,
            "strikethrough": false,
            "underline": false,
          },
          "text": {
            "content": "world",
            "link": undefined,
          },
          "plain_text": "world",
          "type": "text",
        },
      ],
    },
    "type": "paragraph",
  },
  {
    "heading_2": {
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "plain_text": "heading2",
          "text": {
            "content": "heading2",
            "link": undefined,
          },
          "type": "text",
        },
      ],
    },
    "object": "block",
    "type": "heading_2",
  },
  {
    "object": "block",
    "to_do": {
      "checked": true,
      "children": undefined,
      "rich_text": [
        {
          "annotations": {
            "bold": false,
            "code": false,
            "color": "default",
            "italic": false,
            "strikethrough": false,
            "underline": false,
          },
          "text": {
            "content": "todo",
            "link": undefined,
          },
          "plain_text": "todo",
          "type": "text",
        },
      ],
    },
    "type": "to_do",
  },
];