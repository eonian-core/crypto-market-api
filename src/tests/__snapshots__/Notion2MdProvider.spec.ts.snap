// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notion2MdProvider toMd method should convert blocks to markdown correctly 1`] = `
{
  "parent": "
hello _world_


## heading2

- [x] todo
",
}
`;
